pkg/batchmapper/service.go

```diff
@@ -0,0 +1,90 @@
+package batchmapper
+
+import (
+	"context"
+	"io"
+	"log"
+
+	"google.golang.org/protobuf/types/known/emptypb"
+
+	batchmappb "github.com/numaproj/numaflow-go/pkg/apis/proto/batchmap/v1"
+)
+
+const (
+	uds                   = "unix"
+	address               = "/var/run/numaflow/batchmap.sock"
+	defaultMaxMessageSize = 1024 * 1024 * 64
+	serverInfoFilePath    = "/var/run/numaflow/batchmapper-server-info"
+)
+
+// Service implements the proto gen server interface and contains the map operation
+// handler.
+type Service struct {
+	batchmappb.UnimplementedBatchMapServer
+	BatchMapper BatchMapper
+}
+
+// IsReady returns true to indicate the gRPC connection is ready.
+func (fs *Service) IsReady(context.Context, *emptypb.Empty) (*batchmappb.ReadyResponse, error) {
+	return &batchmappb.ReadyResponse{Ready: true}, nil
+}
+
+// BatchMapFn applies a user defined function to a stream of request element and streams back the responses for them.
+func (fs *Service) BatchMapFn(stream batchmappb.BatchMap_BatchMapFnServer) error {
+	ctx := stream.Context()
+
+	// As the BatchMap interface expects a list of request elements
+	// we read all the requests coming on the stream and keep appending them together
+	// and then finally send the array for processing once all the messages on the stream have been read.
+	datums := make([]Datum, 0)
+	for {
+		d, err := stream.Recv()
+		// if we see EOF on the stream we do not have any more messages coming up
+		if err == io.EOF {
+			break
+		}
+		if err != nil {
+			log.Println("BatchMapFn: Got an error while recv() on stream", err)
+			return err
+		}
+		var hd = NewHandlerDatum(d.GetValue(), d.GetEventTime().AsTime(), d.GetWatermark().AsTime(), d.GetHeaders(), d.GetId(), d.GetKeys())
+		datums = append(datums, hd)
+	}
+
+	// Apply the user BatchMap implementation function
+	responses := fs.BatchMapper.BatchMap(ctx, datums)
+
+	// If the number of responses received does not align with the request batch size,
+	// we will not be able to process the data correctly.
+	// This should be marked as an error and shown to the user.
+	if len(responses.Items()) != len(datums) {
+		errMsg := "batchMapFn: mismatch between length of batch requests and responses"
+		log.Panic(errMsg)
+	}
```

Comment:
is it possible to avoid panic?

Reply:
We can choose either to panic or send an error back(numa will be restarted in both cases), based on an earlier discussion with @vigith we thought that having a panic here would be a clear indication for the user that there is an issue. And as this in a mismatch in the implementation from the user side for not providing equal number of responses, the reason for restarts and error gives better clarity there. 
But this is our decision on how do we want to keep the behavior in such scenarios.
@vigith @whynowy Any preferences?